crates/namada/src/ledger/protocol/mod.rs

```diff
@@ -600,18 +652,130 @@ where
                 "Transfer of tx fee cannot be applied to due to fee overflow. \
                  This shouldn't happen."
             );
+            shell_params.state.write_log_mut().drop_batch();
 
             Err(Error::FeeError(format!("{}", e)))
         }
     }
 }
 
-/// Transfer `token` from `src` to `dest`. Returns an `Err` if `src` has
-/// insufficient balance or if the transfer the `dest` would overflow (This can
-/// only happen if the total supply doesn't fit in `token::Amount`). Contrary to
-/// `crate::token::transfer` this function updates the tx write log and
-/// not the block write log.
-fn token_transfer<WLS>(
+fn try_masp_fee_payment<S, D, H, CA>(
+    ShellParams {
+        tx_gas_meter,
+        state,
+        vp_wasm_cache,
+        tx_wasm_cache,
+    }: &mut ShellParams<'_, S, D, H, CA>,
+    tx: &Tx,
+    tx_index: &TxIndex,
+) -> Result<Option<(BatchedTxResult, Hash)>>
+where
+    S: State<D = D, H = H>
+        + StorageRead
+        + Read<Err = namada_state::Error>
+        + Sync,
+    D: 'static + DB + for<'iter> DBIter<'iter> + Sync,
+    H: 'static + StorageHasher + Sync,
+    CA: 'static + WasmCacheAccess + Sync,
+{
+    // The fee payment is subject to a gas limit imposed by a protocol
+    // parameter. Here we instantiate a custom gas meter for this step and
+    // initialize it with the already consumed gas. The gas limit should
+    // actually be the lowest between the protocol parameter and the actual gas
+    // limit of the transaction
```

Comment:
Being the gas limit it's the maximum amount of gas the tx can use but here we make sure that we pick the minimum between the protocol imposed gas limit and the one declared by the transaction, so that if the protocol limit is for example 20k and the transaction sets a limit of 15k we pick the latter (otherwise the transaction would exceed its own limit and won't pay for the whole gas used)

Reply:
Yes, and it is this picking the latter that I would argue makes this a _maximum_ gas limit, since the gas limit can never exceed it